### PR TITLE
Even more profiler fixes/improvements

### PR DESCRIPTION
**Pull request by alexrp**
Decided to combine my other PRs into one to avoid merge conflicts.

### CODE REVIEW

**Comment by alexrp** [comment]
@kumpera the following commits are new since you last reviewed this PR:

* https://github.com/mono/mono/pull/3481/commits/bdd9db1ed1d43e508212e95920ad753887c44c14
* https://github.com/mono/mono/pull/3481/commits/41dc6f631c801a0ef0fcb54c216598b9e2488b9b
* https://github.com/mono/mono/pull/3481/commits/7240d72222c772faf8bea1b88e6acbabe27ab255
* https://github.com/mono/mono/pull/3481/commits/138adfc043af2d4bf8a28fc5ec8fc4dd4970d6d1
* https://github.com/mono/mono/pull/3481/commits/0c0e3702a4ddacec403d38dba7952f2cc617a477
* https://github.com/mono/mono/pull/3481/commits/8310af22304a306bbe9cbc5e9650d997ee227453
* https://github.com/mono/mono/pull/3481/commits/16ea17a5e124ae585502bba633755cb289d5fc3f
* https://github.com/mono/mono/pull/3481/commits/39951e43ef2d1be3ad6df68bd59d9ad29b724ca3
* https://github.com/mono/mono/pull/3481/commits/9a142460536fa82be253bf0815aae8da0bda28bb

File: mono/profiler/proflog.c [hunk 1]
```diff
@@ -1507,12 +1477,11 @@ emit_bt (MonoProfiler *prof, LogBuffer *logbuffer, FrameData *data)
 	 */
 	if (data->count > num_frames)
 		printf ("bad num frames: %d\n", data->count);
-	emit_value (logbuffer, 0); /* flags */
 	emit_value (logbuffer, data->count);
 	//if (*p != data.count) {
 	//	printf ("bad num frames enc at %d: %d -> %d\n", count, data.count, *p); printf ("frames end: %p->%p\n", p, logbuffer->cursor); exit(0);}
 	while (data->count) {
-		emit_method_as_ptr (prof, logbuffer, data->methods [--data->count]);
+		emit_method (prof, logbuffer, data->methods [--data->count]);
```
Comment (kumpera) [inline mono/profiler/proflog.c]: You didn't document that on the backtrace format.

File: mono/profiler/proflog.c [hunk 2]
```diff
@@ -1614,7 +1582,7 @@ gc_roots (MonoProfiler *prof, int num, void **objects, int *root_types, uintptr_
 
 	for (int i = 0; i < num; ++i) {
 		emit_obj (logbuffer, objects [i]);
-		emit_value (logbuffer, root_types [i]);
+		emit_byte (logbuffer, root_types [i]);
```
Comment (kumpera) [inline mono/profiler/proflog.c]: This is a correction to the spec. We should mention that on the history.

File: mono/profiler/proflog.c [hunk 3]
```diff
@@ -2154,7 +2108,7 @@ clause_exc (MonoProfiler *prof, MonoMethod *method, int clause_type, int clause_
 	LogBuffer *logbuffer = ensure_logbuf (
 		EVENT_SIZE /* event */ +
 		LEB128_SIZE /* time */ +
-		LEB128_SIZE /* clause type */ +
+		BYTE_SIZE /* clause type */ +
```
Comment (kumpera) [inline mono/profiler/proflog.c]: TYPE_EXCEPTION is not documented at all.

File: mono/metadata/profiler.h [hunk 4]
```diff
@@ -167,6 +168,9 @@ typedef void (*MonoProfileGCResizeFunc)   (MonoProfiler *prof, int64_t new_size)
 typedef void (*MonoProfileGCHandleFunc)   (MonoProfiler *prof, int op, int type, uintptr_t handle, MonoObject *obj);
 typedef void (*MonoProfileGCRootFunc)     (MonoProfiler *prof, int num_roots, void **objects, int *root_types, uintptr_t *extra_info);
 
+typedef void (*MonoProfileGCFinalizeFunc)  (MonoProfiler *prof);
```
Comment (kumpera) [inline mono/metadata/profiler.h]: This will need to go to the release notes.

File: mono/metadata/mono-gc.h [hunk 5]
```diff
@@ -50,6 +50,8 @@ MONO_API int    mono_gc_get_generation  (MonoObject *object);
 MONO_API int    mono_gc_collection_count (int generation);
 MONO_API int64_t mono_gc_get_used_size   (void);
 MONO_API int64_t mono_gc_get_heap_size   (void);
+MONO_API MonoBoolean mono_gc_pending_finalizers (void);
```
Comment (kumpera) [inline mono/metadata/mono-gc.h]: If we're exposing those on public headers, they must be documented.

File: mono/profiler/decode.c [hunk 6]
```diff
@@ -2400,6 +2400,21 @@ decode_buffer (ProfContext *ctx)
 					fprintf (outfile, "handle (%s) %u destroyed\n", get_handle_name (htype), handle);
 				if (frames != sframes)
 					free (frames);
+			} else if (subtype == TYPE_GC_FINALIZE_START) {
+				// TODO: Generate a finalizer report based on these events.
```
Comment (kumpera) [inline mono/profiler/decode.c]: I can think of 2 reports we'd care to have.

The format can be the same format as the alloc report, but for finalized objects.
We'd report for all objects or broke down per-collection/snapshot.

Second, we could report expensive finalizers, which ones when over X ms.

File: mono/profiler/proflog.c [hunk 7]
```diff
@@ -1619,6 +1622,74 @@ gc_handle (MonoProfiler *prof, int op, int type, uintptr_t handle, MonoObject *o
...
+
+	LogBuffer *buf = ensure_logbuf (
+		EVENT_SIZE /* event */
+	);
+
+	emit_event (buf, TYPE_GC_FINALIZE_START | TYPE_GC);
+
+	EXIT_LOG;
+
+	process_requests (prof);
```
Comment (kumpera) [inline mono/profiler/proflog.c]: Maybe we should only trigger a heapshot at a deterministic point of finalization? Either at begin or end.

IDK whether triggering at any point would not produce some weird artifacts.

**Comment by kumpera** [comment]
LGTM. All issues I mentioned are not blockers.

**Comment by kumpera** [comment]
LGTM for C8

**Comment by alexrp** [comment]
Will address documentation concerns in a follow-up PR.

**Pull request merged**